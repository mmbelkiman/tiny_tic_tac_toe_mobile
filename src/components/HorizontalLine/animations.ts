import {Animated} from 'react-native';

export const playOpacityAnimation = (animation: Animated.Value) => {
  Animated.timing(animation, {
    toValue: 0.6,
    duration: 90,
    useNativeDriver: false,
  }).start();
};

export const playHeightAnimation = (
  animation: Animated.Value,
  height: number,
) => {
  Animated.timing(animation, {
    toValue: height,
    duration: 80,
    useNativeDriver: false,
  }).start();
};

export const playShakeAnimation = (animation: Animated.Value) => {
  Animated.sequence([
    Animated.timing(animation, {
      toValue: 1,
      duration: 200,
      useNativeDriver: false,
    }),
    Animated.timing(animation, {
      toValue: 0,
      duration: 200,
      useNativeDriver: false,
    }),
  ]).start();
};

export const getShakeStyle = (animation: Animated.Value) => [
  {
    translateX: animation.interpolate({
      inputRange: [0, 0.2, 0.4, 0.6, 0.8, 1],
      outputRange: [0, -5, 5, -5, 5, 0],
    }),
  },
];
