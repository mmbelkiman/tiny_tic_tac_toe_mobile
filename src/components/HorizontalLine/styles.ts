import {StyleSheet} from 'react-native';

export default StyleSheet.create({
  line: {
    position: 'absolute',
    width: '90%',
    borderRadius: 10,
    alignSelf: 'center',
  },
});
