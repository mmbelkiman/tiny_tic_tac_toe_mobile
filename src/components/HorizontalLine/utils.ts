import {Animated} from 'react-native';
import {PlayerMark} from '../../common/types';
import {getPlayerColor} from '../../common/utils';
import styles from './styles';
import {getShakeStyle} from './animations';

export const getLineStyle = (
  winner: PlayerMark,
  positionY: number,
  heightAnimation: Animated.Value,
  opacityAnimation: Animated.Value,
  shakeAnimation: Animated.Value,
) => [
  styles.line,
  {
    backgroundColor: getPlayerColor(winner),
    height: heightAnimation,
    top: positionY,
    opacity: opacityAnimation,
  },
  getShakeStyle(shakeAnimation),
];
