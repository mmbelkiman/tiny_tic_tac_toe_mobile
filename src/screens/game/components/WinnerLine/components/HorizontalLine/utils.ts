import {Animated} from 'react-native';
import styles from './styles';

export const getShakeStyle = (animation: Animated.Value) => ({
  transform: [
    {
      translateX: animation.interpolate({
        inputRange: [0, 0.2, 0.4, 0.6, 0.8, 1],
        outputRange: [0, -5, 5, -5, 5, 0],
      }),
    },
  ],
});

export const getLineStyle = (
  color: string,
  positionY: number,
  heightAnimation: Animated.Value,
  opacityAnimation: Animated.Value,
  shakeAnimation: Animated.Value,
) => [
  styles.line,
  {
    backgroundColor: color,
    height: heightAnimation,
    top: positionY,
    opacity: opacityAnimation,
  },
  getShakeStyle(shakeAnimation),
];
