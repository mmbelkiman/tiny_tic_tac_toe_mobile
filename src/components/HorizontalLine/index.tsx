import React, {useEffect, useRef} from 'react';
import {Animated} from 'react-native';
import {PlayerMark} from '@/common/types';
import {
  playOpacityAnimation,
  playShakeAnimation,
  playSizeAnimation,
} from '@/common/animations';
import {getLineStyle} from './utils';

interface HorizontalLineProps {
  height: number;
  positionY: number;
  winner: PlayerMark;
  visible: boolean;
}

const HorizontalLine: React.FC<HorizontalLineProps> = ({
  height,
  positionY,
  winner,
  visible,
}) => {
  const opacityAnimation = useRef(new Animated.Value(0)).current;
  const heightAnimation = useRef(new Animated.Value(0)).current;
  const shakeAnimation = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    if (visible && winner) {
      playOpacityAnimation(opacityAnimation);
      playSizeAnimation(heightAnimation, height);
      playShakeAnimation(shakeAnimation);
    }
  }, [
    height,
    heightAnimation,
    opacityAnimation,
    shakeAnimation,
    visible,
    winner,
  ]);

  if (!visible || winner === null) {
    return null;
  }

  return (
    <Animated.View
      style={getLineStyle(
        winner,
        positionY,
        heightAnimation,
        opacityAnimation,
        shakeAnimation,
      )}
    />
  );
};

export default HorizontalLine;
