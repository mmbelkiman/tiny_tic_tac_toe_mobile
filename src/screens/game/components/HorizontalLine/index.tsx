import React, {useEffect, useRef} from 'react';
import {Animated} from 'react-native';
import {
  playOpacityAnimation,
  playShakeAnimation,
  playSizeAnimation,
} from '@/common/animations';
import {getLineStyle} from './utils';
import {PlayerMark} from '@/screens/game/common/types';

interface HorizontalLineProps {
  height: number;
  positionY: number;
  winner: PlayerMark;
}

const HorizontalLine: React.FC<HorizontalLineProps> = ({
  height,
  positionY,
  winner,
}) => {
  const opacityAnimation = useRef(new Animated.Value(0)).current;
  const heightAnimation = useRef(new Animated.Value(0)).current;
  const shakeAnimation = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    if (winner !== '') {
      playOpacityAnimation(opacityAnimation);
      playSizeAnimation(heightAnimation, height);
      playShakeAnimation(shakeAnimation);
    }
  }, [height, heightAnimation, opacityAnimation, shakeAnimation, winner]);

  if (winner === '') {
    return null;
  }

  return (
    <Animated.View
      style={getLineStyle(
        winner,
        positionY,
        heightAnimation,
        opacityAnimation,
        shakeAnimation,
      )}
    />
  );
};

export default HorizontalLine;
