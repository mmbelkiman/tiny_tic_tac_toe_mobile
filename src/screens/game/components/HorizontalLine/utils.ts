import {Animated} from 'react-native';
import styles from './styles';
import {PlayerMark} from '@game/common/types';
import {getPlayerColor} from '@game/common/utils';

export const getShakeStyle = (animation: Animated.Value) => ({
  transform: [
    {
      translateX: animation.interpolate({
        inputRange: [0, 0.2, 0.4, 0.6, 0.8, 1],
        outputRange: [0, -5, 5, -5, 5, 0],
      }),
    },
  ],
});

export const getLineStyle = (
  winner: PlayerMark,
  positionY: number,
  heightAnimation: Animated.Value,
  opacityAnimation: Animated.Value,
  shakeAnimation: Animated.Value,
) => [
  styles.line,
  {
    backgroundColor: getPlayerColor(winner),
    height: heightAnimation,
    top: positionY,
    opacity: opacityAnimation,
  },
  getShakeStyle(shakeAnimation),
];
