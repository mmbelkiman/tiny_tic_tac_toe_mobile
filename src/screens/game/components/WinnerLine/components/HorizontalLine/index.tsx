import React from 'react';
import {Animated} from 'react-native';
import {getLineStyle} from './utils';

interface HorizontalLineProps {
  positionY: number;
  color: string;
  heightAnimation: Animated.Value;
  opacityAnimation: Animated.Value;
  shakeAnimation: Animated.Value;
  testID?: string | number;
}

const HorizontalLine: React.FC<HorizontalLineProps> = ({
  positionY,
  color,
  heightAnimation,
  opacityAnimation,
  shakeAnimation,
  testID = '',
}) => {
  return (
    <Animated.View
      testID={`HorizontalLine${testID}`}
      style={getLineStyle(
        color,
        positionY,
        heightAnimation,
        opacityAnimation,
        shakeAnimation,
      )}
    />
  );
};

export default HorizontalLine;
