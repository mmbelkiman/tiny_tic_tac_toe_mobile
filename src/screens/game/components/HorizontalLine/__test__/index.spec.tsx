import React from 'react';
import renderer from 'react-test-renderer';
import HorizontalLine from '../index';

jest.mock('@/common/animations', () => ({
  playOpacityAnimation: jest.fn(),
  playSizeAnimation: jest.fn(),
  playShakeAnimation: jest.fn(),
}));

describe('HorizontalLine Component', () => {
  it('should match the snapshot with a winner', () => {
    const tree = renderer
      .create(<HorizontalLine height={100} positionY={50} winner="X" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match the snapshot without a winner', () => {
    const tree = renderer
      .create(<HorizontalLine height={100} positionY={50} winner="" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
