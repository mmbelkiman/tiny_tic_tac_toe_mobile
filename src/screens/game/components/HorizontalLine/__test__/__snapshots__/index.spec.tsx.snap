// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizontalLine Component should match the snapshot with a winner 1`] = `
<View
  collapsable={false}
  style={
    {
      "alignSelf": "center",
      "backgroundColor": "#D80E0D",
      "borderRadius": 10,
      "height": 0,
      "opacity": 0,
      "position": "absolute",
      "top": 50,
      "transform": [
        {
          "translateX": 0,
        },
      ],
      "width": "90%",
    }
  }
/>
`;

exports[`HorizontalLine Component should match the snapshot without a winner 1`] = `null`;
